lib/instrumentation/modules/kafkajs.js

```diff
@@ -0,0 +1,472 @@
+      const parentSpan =
+        span || runContext.currSpan() || runContext.currTransaction();
+
+      if (parentSpan) {
+        record.messages.forEach((msg) => {
+          const newHeaders = Object.assign({}, msg.headers);
+          parentSpan.propagateTraceContextHeaders(
+            newHeaders,
+            function (carrier, name, value) {
+              if (name.startsWith('elastic-')) {
+                return;
+              }
+              carrier[name] = value;
+            },
+          );
+          msg.headers = newHeaders;
+        });
+      }
+
+      if (!span) {
+        return origSend.apply(this, arguments);
+      }
+
+      // We do not add headers or body because:
+      // - `record.messages` is a list
+      // - spec says is for transactions (https://github.com/elastic/apm/blob/main/specs/agents/tracing-instrumentation-messaging.md#transaction-context-fields)
+      span.setMessageContext({ queue: { name: topic } });
+
+      const service = {
+        resource: `${SUBTYPE}/${topic}`,
+        type: SUBTYPE,
+        name: topic,
+      };
+
+      span._setDestinationContext({ service });
+
+      let result, err;
+      try {
+        result = await origSend.apply(this, arguments);
+      } catch (ex) {
+        // Save the error for use in `finally` below, but re-throw it to
+        // not impact code flow.
+        err = ex;
+        throw ex;
+      } finally {
+        span.setOutcome(
+          err ? constants.OUTCOME_FAILURE : constants.OUTCOME_SUCCESS,
+        );
+        span.end();
+      }
+
+      return result;
+    };
+  }
+
+  /**
+   * Returns the patched version of `Producer.sendBatch` which
+   * - creates an exit span for the operation
+   * - propagates trace context via message headers
+   *
+   * @param {Producer['sendBatch']} origSendBatch
+   * @returns {Producer['sendBatch']}
+   */
+  function wrapProducerSendBatch(origSendBatch) {
+    return async function (batch) {
+      let span;
+      let shouldIgnoreBatch = true;
+      const messages = batch.topicMessages || [];
+      const topics = new Set();
+
+      // Remove possible topic duplications
+      for (const msg of messages) {
+        topics.add(msg.topic);
+      }
+
+      // Check if the batch needs to be ignored
+      for (const t of topics) {
+        if (!shouldIgnoreTopic(t, config)) {
+          shouldIgnoreBatch = false;
+          break;
+        }
+      }
+
+      if (!shouldIgnoreBatch) {
+        span = ins.createSpan(
+          `${NAME} send messages batch`,
+          TYPE,
+          SUBTYPE,
+          'send',
+          { exitSpan: true },
+        );
+      }
+
+      // W3C trace-context propagation.
+      const runContext = ins.currRunContext();
+      const parentSpan =
+        span || runContext.currSpan() || runContext.currTransaction();
+
+      if (parentSpan && batch.topicMessages) {
+        batch.topicMessages.forEach((topicMessage) => {
+          topicMessage.messages.forEach((msg) => {
```

Comment:
Should we avoid propagating trace-context on messages in a topic that we should ignore?

Reply:
Nevermind, I am wrong. This should still propagate the trace-context.